src/Hosting/SystemdServices/src/SystemdNotifier.cs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Net.Sockets;
+
+namespace Microsoft.Extensions.Hosting.SystemdServices
+{
+    public class SystemdNotifier : ISystemdNotifier
+    {
+        const string NOTIFY_SOCKET = "NOTIFY_SOCKET";
+
+        private readonly string _socketPath;
+
+        public SystemdNotifier() :
+            this(GetNotifySocketPath())
+        { }
+
+        // For testing
+        internal SystemdNotifier(string socketPath)
+        {
+            _socketPath = socketPath;
+        }
+
+        /// <inheritdoc />
+        public bool IsEnabled => _socketPath != null;
+
+        /// <inheritdoc />
+        public void Notify(ServiceState state, ServiceState[] states)
+        {
+            if (!IsEnabled)
+            {
+                return;
+            }
+
+            byte[] data;
+            if (states.Length == 0)
+            {
+                data = Encoding.UTF8.GetBytes(state.ToString());
+            }
+            else
+            {
+                var ms = new MemoryStream();
+                AppendState(ms, state);
+                for (int i = 0; i < states.Length; i++)
+                {
+                    ms.WriteByte((byte)'\n');
+                    AppendState(ms, states[i]);
+                }
+                data = ms.ToArray();
+            }
+
+            using (var socket = new Socket(AddressFamily.Unix, SocketType.Dgram, ProtocolType.Unspecified))
+            {
+                var endPoint = new UnixDomainSocketEndPoint(_socketPath);
+                socket.Connect(endPoint);
+
+                int rv = socket.Send(data);
```

Comment:
1. Async?
2. Errors?

Reply:
I thought about async too, but it's called in a sync context, so there will be waiting anywhere.
Is `socket.SendAsync(...).GetAwaiter().GetResult()` a better option here?

Reply:
It's currently called in a single place but this is a general purpose interface. We may need both sync and async versions.

Reply:
> Is `socket.SendAsync(...).GetAwaiter().GetResult()` a better option here?

Definitely not. I'd make this a `NotifyAsync` method and allow callers to decide what to do. For example, a caller could decide to fire-and-forget instead of blocking.

Reply:
We can treat this as a sync API. The buffer in the kernel is far larger than what we are trying to send.

Reply:
Good to know. Can you add a comment for that fact please?

Reply:
> We can treat this as a sync API. The buffer in the kernel is far larger than what we are trying to send.

Hrm, we generally avoid operating under assumptions like that, but I see your point. Thoughts @davidfowl ?